### PR TITLE
8331189: Implementation of Scoped Values (Third Preview)

### PR DESCRIPTION
**Pull request by AlanBateman**
JEP 481 proposes Scoped Values to continue to preview in JDK 23 with one change. The type of the operation parameter of the callWhere method is changed to a new functional interface to avoid having the API throw Exception. With that change, the getWhere (and the corresponding method on Carrier) are no longer needed. The functional interface is not proposed for j.u.function at this time.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8331190](https://bugs.openjdk.org/browse/JDK-8331190) to be approved

### Issues
 * [JDK-8331189](https://bugs.openjdk.org/browse/JDK-8331189): Implementation of Scoped Values (Third Preview) (**Enhancement** - P4)
 * [JDK-8331190](https://bugs.openjdk.org/browse/JDK-8331190): Implementation of Scoped Values (Third Preview) (**CSR**)

### Reviewers
 * [Andrew Haley](https://openjdk.org/census#aph) (@theRealAph - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Maurizio Cimadamore](https://openjdk.org/census#mcimadamore) (@mcimadamore - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19136/head:pull/19136` \
`$ git checkout pull/19136`

Update a local copy of the PR: \
`$ git checkout pull/19136` \
`$ git pull https://git.openjdk.org/jdk.git pull/19136/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19136`

View PR using the GUI difftool: \
`$ git pr show -t 19136`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19136.diff">https://git.openjdk.org/jdk/pull/19136.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19136#issuecomment-2122066992)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 1]
```diff
@@ -532,15 +488,33 @@ private void runWith(Snapshot newSnapshot, Runnable op) {
...
+    /**
+     * An operation that returns a result and may throw an exception.
+     *
+     * @param <T> result type of the operation
+     * @param <X> type of the exception thrown by the operation
+     * @since 23
+     */
+    @PreviewFeature(feature = PreviewFeature.Feature.SCOPED_VALUES)
+    @FunctionalInterface
+    public interface CallableOp<T, X extends Throwable> {
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/ScopedValue.java:500]: I feel like that a functional interface such as this one should be promoted to a top‑level type in some common package by the time this feature exits preview.

**Review by theRealAph — APPROVED** [review]
Thanks. That all looks reasonable enough, and gets us to the status quo ante + `ScopedValue.CallableOp`,  which is way better than what we have right now.

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 2]
```diff
@@ -532,6 +490,24 @@ private void runWith(Snapshot newSnapshot, Runnable op) {
         }
     }
 
+    /**
+     * An operation that returns a result and may throw an exception.
```
Comment (jaikiran) [inline src/java.base/share/classes/java/lang/ScopedValue.java:494]: Nit - should it be "or may throw an exception"?

**Review by jaikiran — APPROVED** [review]
The changes look OK to me.

**Review by mcimadamore — APPROVED** [review]
Changes look good. I like how the new functional interface makes the API seem smaller.

**Pull request closed without merging**